pkg/client/client_test.go

```diff
@@ -2304,3 +2308,60 @@ func assertErrorsMatch(t *testing.T, want, got []error) {
 		}
 	}
 }
+
+func TestPollUntil(t *testing.T) {
+	testPoll := func(ctx context.Context, condition wait.ConditionWithContextFunc, err *error) error {
+		return Poll(ctx, condition, WithPollInterval(100*time.Millisecond), WithPollTimeout(150*time.Millisecond), WithLastError(err))
+	}
+
+	// condition is met.
+	var lastErr1 error
+	err := testPoll(context.Background(), func(ctx context.Context) (bool, error) {
+		return true, nil
+	}, &lastErr1)
+	require.NoError(t, err)
+
+	// condition is met, lastErr isn't used.
+	err = testPoll(context.Background(), func(ctx context.Context) (bool, error) {
+		return true, nil
+	}, nil)
+	require.NoError(t, err)
+
+	// condition is eventually met.
+	var lastErr2 error
+	err = testPoll(context.Background(), func(ctx context.Context) (bool, error) {
+		lastErr2 = fmt.Errorf("INSIDE ERROR")
+		return true, nil
+	}, &lastErr2)
+	require.NoError(t, err)
+
+	// condition returns an error before timeout.
+	var lastErr3 error
+	err = testPoll(context.Background(), func(ctx context.Context) (bool, error) {
+		insideError := fmt.Errorf("INSIDE ERROR")
+		lastErr3 = insideError
+		return false, insideError
+	}, &lastErr3)
+	require.ErrorContains(t, err, "INSIDE ERROR: lastError: INSIDE ERROR")
```

Comment:
(nit) it might more readable/plausible to have 2 different errors (e.g. `INSIDE ERROR` and `OUTSIDE ERROR`)

Reply:
Do you suggest we call this error "OUTSIDE ERROR"?

(Note that in these "rare" situations, the error returned by `PollUntilContextTimeout` and `lastError` will be the same, and I don't think the logic (I haven't thought of any) to avoid this duplication is worth it)

Reply:
I think a simple comparison should do it, pushed a new version.